snapshot/diffapply_unix.go

```diff
@@ -24,240 +26,353 @@ import (
+			} else if info.Kind == snapshots.KindCommitted {
+				lowerMntable, err = sn.View(ctx, identity.NewID(), diff.Lower)
+				if err != nil {
+					return snapshots.Usage{}, errors.Wrapf(err, "failed to mount lower snapshot view %s", diff.Lower)
+				}
+			} else {
+				lowerMntable, err = sn.Mounts(ctx, diff.Lower)
+				if err != nil {
+					return snapshots.Usage{}, errors.Wrapf(err, "failed to mount lower snapshot %s", diff.Lower)
+				}
+			}
+		}
+		var upperMntable Mountable
+		if diff.Upper != "" {
+			if info, err := sn.Stat(ctx, diff.Upper); err != nil {
+				return snapshots.Usage{}, errors.Wrapf(err, "failed to stat upper snapshot %s", diff.Upper)
+			} else if info.Kind == snapshots.KindCommitted {
+				upperMntable, err = sn.View(ctx, identity.NewID(), diff.Upper)
+				if err != nil {
+					return snapshots.Usage{}, errors.Wrapf(err, "failed to mount upper snapshot view %s", diff.Upper)
+				}
+			} else {
+				upperMntable, err = sn.Mounts(ctx, diff.Upper)
+				if err != nil {
+					return snapshots.Usage{}, errors.Wrapf(err, "failed to mount upper snapshot %s", diff.Upper)
+				}
+			}
+		} else {
+			// create an empty view
+			upperMntable, err = sn.View(ctx, identity.NewID(), "")
+			if err != nil {
+				return snapshots.Usage{}, errors.Wrapf(err, "failed to mount empty upper snapshot view %s", diff.Upper)
+			}
+		}
+		d, err := differFor(lowerMntable, upperMntable)
 		if err != nil {
-			return err
+			return snapshots.Usage{}, errors.Wrapf(err, "failed to create differ")
+		}
+		defer func() {
+			rerr = multierror.Append(rerr, d.Release()).ErrorOrNil()
+		}()
+		if err := d.HandleChanges(ctx, a.Apply); err != nil {
+			return snapshots.Usage{}, errors.Wrapf(err, "failed to handle changes")
 		}
-		lowerMounts = mounts
-		defer unmountLower()
 	}
 
-	var upperMounts []mount.Mount
-	if upperMountable != nil {
-		mounts, unmountUpper, err := upperMountable.Mount()
-		if err != nil {
-			return err
-		}
-		upperMounts = mounts
-		defer unmountUpper()
+	if err := a.Flush(); err != nil {
+		return snapshots.Usage{}, errors.Wrapf(err, "failed to flush changes")
 	}
+	return a.Usage()
+}
 
-	lowerMnter := LocalMounterWithMounts(lowerMounts)
-	lowerView, err := lowerMnter.Mount()
-	if err != nil {
-		return err
+type change struct {
+	kind    fs.ChangeKind
+	subPath string
+	srcPath string
+	srcStat *syscall.Stat_t
+	// linkSubPath is set to a subPath of a previous change from the same
+	// differ instance that is a hardlink to this one, if any.
+	linkSubPath string
+}
+
+type changeApply struct {
+	*change
+	dstPath string
+	dstStat *syscall.Stat_t
+}
+
+type inode struct {
+	ino uint64
+	dev uint64
+}
+
+func statInode(stat *syscall.Stat_t) inode {
+	if stat == nil {
+		return inode{}
+	}
+	return inode{
+		ino: stat.Ino,
+		dev: stat.Dev,
 	}
-	defer lowerMnter.Unmount()
+}
 
-	upperMnter := LocalMounterWithMounts(upperMounts)
-	upperView, err := upperMnter.Mount()
-	if err != nil {
-		return err
+type applier struct {
```

Comment:
As commented in earlier PR I think this would be better as a separate pkg in some follow up that would hopefully lead to deduplication between this and similar tar implementations in containerd.